### PR TITLE
Improve HTTP “framework” to support generated documentations

### PR DESCRIPTION
**Pull request by schubev**
Improve the HTTP “framework” in various ways to ease documentation.

- Add features to document routes within the code.
- Decorators can affect the documentation.
- Add the MethodMuxer to select handlers based on HTTP method.
- Improve build process with a build number used in the backend ID.
- Remove request time from route logging context.
- Replace route registrations with new system.
- Respond to OPTIONS requests with route documentation.
- Expose API documentation on /docs.
- Unpublish ErrorBody.
- More correctly handle CORS OPTIONS requests.

### CODE REVIEW

File: routes/decoratorBackendId.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+//   limitations under the License.
+
+package routes
+
+import (
+	"net/http"
+)
+
+// BackendId is a decorator which adds a backend ID to incoming requests. It
+// adds it to tthe response in the `X-Backend-ID` HTTP header.
```
Comment (giubil) [inline routes/decoratorBackendId.go]: `tthe`

File: routes/decoratorCors.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	AllowOrigin      []string
+	AllowHeaders     []string
+	AllowCredentials bool
+}
+
+func (c Cors) Decorate(h Handler) Handler {
+	mm := h.methods
+	ms := make([]string, 0, len(mm)+1)
+	ms = append(ms, http.MethodOptions)
+	for m, v := range mm {
```
Comment (giubil) [inline routes/decoratorCors.go]: `mm`, `m` and `v`are not explicit enough

File: routes/decoratorDocumentation.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+//   limitations under the License.
+
+package routes
+
+// Documentation decorates a handler to document it. Summary, Description and
+// Tags will be set on the documentation if not zero.
+type Documentation HandlerDocumentationBody
+
+func (d Documentation) Decorate(h Handler) Handler {
+	n := h.Documentation
```
Comment (giubil) [inline routes/decoratorDocumentation.go]: Maybe you could use something a bit more specific than `n` as a variable name

File: routes/decoratorQueryArgs.go [hunk 4]
```diff
@@ -26,14 +26,16 @@ const (
...
 )
 
 type (
 	// QueryArgInt denotes an int query argument. It fulfills the QueryParser
 	// interface.
 	QueryArgInt struct{}
 
-	// QueryArgUint denotes an uint query argument. It fulfills the QueryParser
+	// QueryArgUint denotes a uint query argument. It fulfills the QueryParser
 	// interface.
```
Comment (giubil) [inline routes/decoratorQueryArgs.go:39]: Why did you change `an` to `a` here ? `uint` starts with a vowel sound so shouldn't it be `an` ?

File: routes/decoratorRequestContentType_test.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+package routes
+
+import (
+	"encoding/json"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+const testDocumentedHandlerWithRequestContentTypeExpected = `{"summary":"Get yourself some foo","description":"The route gives you some foo.","tags":{"require:contenttype":["application/json","application/csv"]}}`
```
Comment (giubil) [inline routes/decoratorRequestContentType_test.go]: This line is 214 characters long, you could pretty print the json instead of having it on one line

File: routes/decoratorRequestTime.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+import (
+	"net/http"
+	"time"
+)
+
+// RequestTime is a decorator which adds a request time to incoming requests.
+// It stores it in the request context and sets the logger up to log it.
+type RequestTime struct{}
+
+func (d RequestTime) Decorate(h Handler) Handler {
```
Comment (giubil) [inline routes/decoratorRequestTime.go]: This method is not tested

File: routes/decoratorRequestId.go [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+	"github.com/satori/go.uuid"
+)
+
+// RequestId is a decorator which adds a request ID to incoming requests.  It
+// stores it in the request context, sets the logger up to log it, and adds it
+// to the response in the `X-Request-ID` HTTP header.
+type RequestId struct{}
+
+func (ri RequestId) Decorate(h Handler) Handler {
```
Comment (giubil) [inline routes/decoratorRequestId.go:28]: This method is not tested

File: routes/decoratorRouteLog.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+	Host      string    `json:"host"`
+	UserAgent []string  `json:"userAgent"`
+	Time      time.Time `json:"time"`
+}
+
+// RouteLog is a decorator which logs any calls to the route, with some data
+// about the request.
+type RouteLog struct{}
+
+func (rl RouteLog) Decorate(h Handler) Handler {
```
Comment (giubil) [inline routes/decoratorRouteLog.go:52]: This function is not tested as well

File: routes/documentation_test.go [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+package routes
+
+import (
+	"encoding/json"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+const testDocumentationExpected = `{"/doc":{"summary":"get the api's documentation","components":{"method:GET":{"summary":"get the api's documentation","description":"Get the api's documentation in structured (JSON) format. This documentation is automatically generated from the definition of the route handlers and thus should always be up to date. The same documentation can be obtained for specific routes using the OPTIONS request on them."}}}}`
```
Comment (giubil) [inline routes/documentation_test.go]: This json should be pretty printed to avoid a long line

File: routes/routes.go [hunk 10]
```diff
@@ -28,61 +30,60 @@ var RegisteredHandlers = make([]RegisteredHandler, 0, 0x40)
...
 type RegisteredHandler struct {
 	Pattern string
-	Handler IntermediateHandler
+	Handler
+}
+
+// type Handler interface {
+// 	Handle(http.ResponseWriter, *http.Request, Arguments) (int, interface{})
+// 	Documentation() HandlerDocumentation
+// }
```
Comment (giubil) [inline routes/routes.go]: Why did you left this commented code in the PR ?

File: routes/routes_test.go [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+	return http.StatusOK, getFooResponse
+}
+
+const postFooResponse = "GIMME YOUR FOO"
+
+func postFoo(r *http.Request, a Arguments) (int, interface{}) {
+	return http.StatusOK, postFooResponse
+}
+
+const testDocumentedHandlerExpected = `{"summary":"Get yourself some foo","description":"The route gives you some foo.","tags":{"foo":["yes"]}}`
```
Comment (giubil) [inline routes/routes_test.go]: You could also pretty print that json to make it easier to read

File: routes/routes_test.go [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+		}
+		rh.Handler.Func(nil, nil, nil)
+		if !handlerRun {
+			t.Error("Handler should have run, hasn't.")
+		}
+	}
+	resetRegisteredHandlers()
+}
+
+const testMethodMuxerDocumentationExpected = `{"summary":"Interacts with foo.","components":{"method:GET":{"summary":"Gets foo."},"method:POST":{"summary":"Posts foo."}}}`
```
Comment (giubil) [inline routes/routes_test.go]: Ditto for this json string

File: users/decorator.go [hunk 13]
```diff
@@ -23,25 +23,40 @@ import (
...
+	return h
+}
+
+func (_ RequireAuthenticatedUser) getFunc(hf routes.HandlerFunc) routes.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request, a routes.Arguments) (int, interface{}) {
 		auth := r.Header["Authorization"]
 		tx := a[db.Transaction].(*sql.Tx)
 		if auth != nil && len(auth) == 1 {
 			tokenString := auth[0]
 			if user, err := testToken(tx, tokenString); err == nil {
```
Comment (giubil) [inline users/decorator.go:49]: Here you check if the `err` is `nil` but you never handle it in the case that it is not `nil`

**Review by giubil — CHANGES_REQUESTED** [review]

File: routes/decoratorQueryArgs.go (hunk 4 above)
Comment (schubev) [reply routes/decoratorQueryArgs.go:39]: It depends on the pronunciation : you would say “an umbrella”, because it starts with [ʌ], while you would say “a union” because it starts with [eə]. I, personnaly, pronounce “uint” [eəInt], hence the “a”.

**Review by giubil — APPROVED** [review]
